engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+// If the resource requires changes, CheckApply returns false regardless of the
+// apply value, true otherwise. Any errors are wrapped and returned.
+func (obj *HetznerVMRes) CheckApply(apply bool) (bool, error) {
+	checkOK := true
+	ctx := context.TODO()
+	// Request up-to-date server info from the API.
+	if err := obj.getServerUpdate(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "getServerUpdate failed")
+	}
+	// Check if the server instance is in the correct state.
+	if c, err := obj.checkApplyServerState(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyServerState failed")
+	} else if !c {
+		checkOK = false
+	}
+	// Skip further steps if not in the intended state.
+	if !apply && !checkOK {
+		return false, nil
+	}
+	// Skip further steps if the server is not running.
+	if obj.State != StateRunning {
+		return checkOK, nil
+	}
+	// Changes in cpu, location and/or image require a server rebuild.
+	if c, err := obj.checkApplyServerRebuild(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyServerRebuild failed")
+	} else if !c {
+		checkOK = false
+	}
+	// Changes in rescue mode can be made while the server is running.
+	if c, err := obj.checkApplyRescueMode(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyRescueMode failed")
+	} else if !c {
+		checkOK = false
+	}
+	// TODO: checkapply SSH keys
+	return checkOK, nil
+}
+
+// checkApplyServerState powers the server on/off, creates a new instance, or
+// deletes the existing one as needed, to ensure the correct state.
+// AllowDeletion provides basic protection against unwanted server deletions.
+func (obj *HetznerVMRes) checkApplyServerState(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerState(apply: %t)", apply)
+	}
+	checkOK := true
+	// Make sure the server exists as intended before further checks.
+	// TODO: add create permission
+	if obj.server == nil {
+		if obj.State == StateAbsent {
+			return true, nil
+		}
+		if !apply {
+			return false, nil
+		}
+		checkOK = false
+		if err := obj.createServer(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "createServer failed")
+		}
+	}
+	// Verify if the server is in a steady state before further steps.
+	// TODO: would it be possible to skip waiters by stacking requests?
+	// TODO: this would complicate the checkapply flow, build tests first!
+	if err := obj.waitFor(ctx, serverInSteadyState); err != nil {
+		return false, errwrap.Wrapf(err, "waitFor(serverInSteadyState) exited early")
+	}
+
+	// If the state has already converged, exit early.
+	stateConverged, err := serverStateConverged(obj)
+	if err != nil {
+		return checkOK, errwrap.Wrapf(err, "confirmStateConvergence failed")
+	}
+	if stateConverged {
+		return checkOK, nil
+	}
+
+	// Request changes if the server is not in the intended state yet.
+	// TODO: add permissions? (poweron, poweroff, delete)
+	checkOK = false
```

Comment:
Yes, but looking back at the code I agree that it might need a clearer rewrite. Maybe a more specific flag like `serverExistsAsIntended` would improve readability? Will take a moment to consider the options. 

As for logic: the checkOk returned on line 316 is 
* `true` if the server did not require creation in the first check, and is already in the desired state (running/off). 
* `false` if the server was absent and required creation, or if it still needs to be pushed to the desired state (running/off).